kubeflow/kubebench/kubebench-job.libsonnet

```diff
@@ -0,0 +1,355 @@
+          },
+        },
+      ];  // secretVols
+      local baseVols = [
+        {
+          name: kubebenchConfigVol,
+          persistentVolumeClaim: {
+            claimName: configPvc,
+          },
+        },
+        {
+          name: kubebenchExpVol,
+          persistentVolumeClaim: {
+            claimName: experimentPvc,
+          },
+        },
+        if dataPvc != "null" then {
+          name: kubebenchDataVol,
+          persistentVolumeClaim: {
+            claimName: dataPvc,
+          },
+        },
+      ];  // baseVols
+
+      local secretVolMnts = [
+        if githubTokenSecret != "null" then {
+          name: kubebenchGithubTokenVol,
+          mountPath: "/secret/github-token",
+        },
+        if gcpCredentialsSecret != "null" then {
+          name: kubebenchGcpCredsVol,
+          mountPath: "/secret/gcp-credentials",
+        },
+      ];  // secretVolMnts
+      local baseVolMnts = [
+        {
+          name: kubebenchConfigVol,
+          mountPath: kubebenchConfigRoot,
+        },
+        {
+          name: kubebenchExpVol,
+          mountPath: kubebenchExpRoot,
+        },
+        if dataPvc != "null" then {
+          name: kubebenchDataVol,
+          mountPath: kubebenchDataRoot,
+        },
+      ];  // baseVolMnts
+
+      local buildStep(name, template, argParams=[]) = {
+        name: name,
+        template: template,
+      } + if std.length(argParams) > 0 then {
+        arguments: {
+          parameters: argParams,
+        },
+      } else {};  // buildStep
+
+      local buildTemplate(stepName, image, command, envVars=[], volMnts=[], inParams=[], outParams=[]) = {
+        name: stepName,
+        container: {
+          command: command,
+          image: image,
+          env: envVars,
+          volumeMounts: volMnts,
+        },
+      } + if std.length(inParams) > 0 then {
+        inputs: {
+          parameters: inParams,
+        },
+      } else {} + if std.length(outParams) > 0 then {
+        outputs: {
+          parameters: outParams,
+        },
+      } else {};  // buildTemplate
+
+      {
+        apiVersion: "argoproj.io/v1alpha1",
+        kind: "Workflow",
+        metadata: {
+          name: name,
+          namespace: namespace,
+        },
+        spec: {
+          [if serviceAccount != "default" then "serviceAccountName"]: serviceAccount,
+          entrypoint: "kubebench-workflow",
+          volumes: secretVols + baseVols,
+          templates: [
+            {
+              name: "kubebench-workflow",
+              steps: [
+                [buildStep("run-configurator", "configurator")],
+                [
+                  buildStep(
+                    "launch-main-job",
+                    "main-job",
+                    argParams=[
+                      {
+                        name: "kf-job-manifest",
+                        value: "{{steps.run-configurator.outputs.parameters.kf-job-manifest}}",
```

Comment:
what does this {{}} mean?

Reply:
there are 2 steps involved here
step 1: run configurator, which generates a manifest of kubeflow job (e.g. tfjob/pytorchjob/etc.)
step 2: run the "main job" using the above generated manifest
the {{}} here is the manifest being passed between the 2 steps

Reply:
I see. Very cool, I didn't know this. Thanks!